### ISSUE TITLE
ipynb reader: application/javascript is currently stripped

### ISSUE THREAD

**Issue by ickc** [jgm/pandoc#7750]
Currently, if there is a `application/javascript` in the `outputs`, pandoc wouldn't included them in the native.
The expectation is to keep them somehow and properly write them in e.g. html output.

Example:

[just-mock.zip](https://github.com/jgm/pandoc/files/7696395/just-mock.zip)

You'll find `just-mock.ipynb` inside accompanied with a few of its output by

```bash
jupyter-nbconvert just-mock.ipynb --to html
pandoc -s -o just-mock-pandoc.html just-mock.ipynb -V header-includes='<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js"></script>' --extract-media=media
pandoc -s -o just-mock.native just-mock.ipynb
```

Focusing on the snippet at

```json
"data": {
      "application/javascript": [
          "some_javascript_here"
      ]
     }
```

pandoc didn't include them from ipynb to native.

While by `nbconvert`, it is transformed as

```html
<div id="a9b8803a-9db2-44ab-b177-f43bef656cd0" class="jp-RenderedJavaScript jp-OutputArea-output " data-mime-type="application/javascript">
<script type="text/javascript">
var element = document.getElementById('a9b8803a-9db2-44ab-b177-f43bef656cd0');
some_javascript_here
</script>
</div>
```

Note that the javascript is inlined with `<script ...</script>` regardless if there's a corresponding html in the same cell.
And some thin code is surrounded to pass the div id to the javascript, with a newly generated id that is not in the original ipynb.

I actually do not understand pandoc enough to know where/how this occur, i.e. if it is the jgm/ipynb issue or here.

Practically, some jupyter extensions uses this technique to show interactive plots. I know that bokeh does this, but not plotly. See a [bokeh example here](https://github.com/ickc/pandoc-ipynb/blob/master/just-bokeh.ipynb).

Another note is that nbconvert has a [sanitize pre-processor](https://github.com/jupyter/nbconvert/blob/ff334777496bb77a4b2bcd3a79c63d8654ca632c/nbconvert/preprocessors/sanitize.py), which strips the javascript from the notebook when output to html. I suppose this is how github preview a ipynb such as the [link above](https://github.com/ickc/pandoc-ipynb/blob/master/just-bokeh.ipynb). So probably we should have an option in pandoc toggling this behavior as well (or add that behavior to some existing flags.)

**Comment by jgm** [comment]
We could render this kind of thing in a raw HTML block.  I'm not sure I understand the `var element =` part above.  Is there a convention that the contents of an `application/javascript` can use the variable `element` and that it will be set in this way?  IS any of this documented anywhere?

**Comment by ickc** [comment]
I don't think it is documented anywhere. I think it is a question for the upstream but I don't know how to start asking.

For example, this is the code in nbconvert adding the `var element`:

https://github.com/jupyter/nbconvert/blob/6c8d9c8f8374f8785967e0e622ad10826228c193/share/jupyter/nbconvert/templates/lab/base.html.j2#L234-L286

`data_javascript`, `data_widget_state`, `data_widget_view` are put inside a script tag. The later 2 has `{{ output.data[datatype] | json_dumps }}`, which seems to mean that *some* json are dumped here, but not the one I reported earlier (I mentioned `application/json` is not contained in the output, and a certain `application/vnd...+json` too.)

(Also, the footer there add `application/vnd.jupyter.widget-state+json`, so at least some `+json` would remains.

It seems like nbformat while providing a spec, is just very general on what is valid, but not on what they mean (i.e. how they should be handled say in html output.) And then they have nbconvert which is a "reference implementation" on how to convert from ipynb to html. But from the look of it, they don't seem to expect "3rd party implementation" as many of the details are left out.

(But I guess the situation is at least better than some other formats that pandoc supports that we can't even look into the source.)

**Comment by ickc** [comment]
There's some documentation from [Embedding Jupyter Widgets in Other Contexts than the Notebook — Jupyter Widgets 8.0.0b0 documentation](https://ipywidgets.readthedocs.io/en/latest/embedding.html?highlight=element#python-interface).
